plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1847,6 +1871,85 @@ private void prepareAlterPartition(HdfsContext hdfsContext, HiveIdentity identit
                     new PartitionWithStatistics(oldPartition.get(), partitionName, oldPartitionStatistics)));
         }
 
+        private void handleFilesForFailedRetryAttempts(HdfsContext hdfsContext, String queryId, PartitionAndMore partitionAndMore)
+        {
+            if (!partitionAndMore.isCleanExtraOutputFilesOnCommit()) {
+                return;
+            }
+            if (!partitionAndMore.hasFileNames()) {
+                return;
+            }
+            deleteFilesForFailedRetryAttempts(hdfsContext, queryId, partitionAndMore.getCurrentLocation(), ImmutableSet.copyOf(partitionAndMore.getFileNames()));
+        }
+
+        private void handleFilesForFailedRetryAttempts(HdfsContext hdfsContext, String queryId, TableAndMore tableAndMore)
+        {
+            if (!tableAndMore.isCleanExtraOutputFilesOnCommit()) {
+                return;
+            }
+            if (tableAndMore.getCurrentLocation().isEmpty()) {
+                return;
+            }
+            if (tableAndMore.getFileNames().isEmpty()) {
+                return;
+            }
+
+            Path tableLocation = tableAndMore.getCurrentLocation().orElseThrow(() -> new IllegalArgumentException("table location not set"));
+            List<String> files = tableAndMore.getFileNames().orElseThrow(() -> new IllegalStateException("file list not set"));
+            deleteFilesForFailedRetryAttempts(hdfsContext, queryId, tableLocation, ImmutableSet.copyOf(files));
+        }
+
+        private void deleteFilesForFailedRetryAttempts(HdfsContext hdfsContext, String queryId, Path path, Set<String> filesToKeep)
+        {
+            List<String> filesToDelete = new LinkedList<>();
+            try {
+                log.debug("Deleting failed attempt files from %s for query %s", path, queryId);
+                FileSystem fileSystem = hdfsEnvironment.getFileSystem(hdfsContext, path);
+                if (!fileSystem.exists(path)) {
+                    throw new TrinoException(HIVE_FILESYSTEM_ERROR, format("Cannot delete failed attempt files from %s for query %s; directory does not exist", path, queryId));
+                }
+
+                // files are written by use flat in single directory so we do not need to list recursively
+                RemoteIterator<LocatedFileStatus> iterator = fileSystem.listFiles(path, false);
+                while (iterator.hasNext()) {
+                    Path file = iterator.next().getPath();
+                    if (isFileCreatedByQuery(file.getName(), queryId) && !filesToKeep.contains(file.getName())) {
```

Comment:
I do not believe there is any documentation. What you see here is strictly specific to Hive connector. 
Engine-wise commit protocol is just a call to one of:
* ConnectorMetadata.finishCreateTable
* ConnectorMetadata.finishInsert
* ConnectorMetadata.finishUpdate
* ConnectorMetadata.finishRefreshMaterializedView
* ConnectorMetadata.finishTableExecute

Each of these methods takes `Collection<Slice> fragments`, which is a collection of connector specific opaque objects returned from calls to `ConnectorPageSink.finish()`.
In case of Hive `HivePageSink.finish()` returns a list of files written by the page sink. 

Depending on various conditions (filesystem in use, configuration, ...) the files are written directly into table/partition directory or to a staging directory. 
In former case commit (finish*) does not have much to do. It just registers table/partition in HMS/Glue (if needed).
In latter case data needs to be moved to final location.